### ISSUE TITLE
When importing pyright infers types of variables from the except clause only

### ISSUE THREAD

**Issue by AmjadHD** [microsoft/pyright#973]
**Description**
Hello, I have two files `mytyping.py` with this content:
```python
try:
    from typing import List
except:
    List = "Hello"
```
and `foo.py` with this content:
```python
from mytyping import List

def f(arg: List[str]):
    return arg
```
pyright complains: `Expected class type but received "Literal['Hello']"`, it infers `List` as `(variable) List: Literal['Hello']`.
if I flip the code in the try/except clauses in `mytyping.py`:
```python
try:
    List = "Hello"
except:
    from typing import List
```
The error disappears (it infers `List` as `(type alias) List: Type[List[_T]]`)
So pyright only considers code in the `except` clause

but when I modify the code in `foo.py` to:
```python
try:
	List = "Hello"
except:
	from typing import List

def f(arg: List[str]):
	return arg
```
pyright infers `List` as `(variable) List: List | Literal['Hello']` and complains (obviously)

**VS Code extension or command-line**
command line tool through https://github.com/sublimelsp/LSP-pyright
pyright 1.1.65 and 1.1.64

**Additional context**
python 3.8
adding a `pyrightconfig.json` with `pythonVersion` does not change anything.

**Comment by erictraut** [comment]
I just responded in gitter, but for completeness, I'll post my response here too:

Can you explain what you are trying to do? I'm very confused by this code. In some code paths (the "try" block) you are assigning a class to the symbol "List" and in another code path (the "except" block) you are assigning a str instance to the same symbol. Then you are using the symbol "List" in a type annotation. Type annotations need to be classes, not instances. The type checker is telling you that you are not using a class type in an annotation.

**Comment by AmjadHD** [comment]
This was only a poc, the real problem i faced was when working with a library that needs to work with two python runtimes 3.3 and 3.8 so it needs to handle types this way https://github.com/sublimelsp/LSP/blob/master/plugin/core/typing.py
this produces errors in all over the code base
![image](https://user-images.githubusercontent.com/33235928/91494885-e7933700-e8b9-11ea-84e1-9bd6310b9035.png)
when i switch the import to `from typing import ...` everything works fine
![image](https://user-images.githubusercontent.com/33235928/91494977-0c87aa00-e8ba-11ea-9063-c0722a5dd42f.png)

**Comment by AmjadHD** [comment]
This is a mix of two "problems", pyright does not like redefining types in the way it was done in the library unlike mypy i think and the second is that pyright always infers the types from the except clause and ignores the `try` even though I'm running python 3.8 and the import in the try clause should work.

**Comment by erictraut** [comment]
I still don't understand the problem you're trying to solve. What feature are you using that is not supported in Python 3.3? Python 3.3 supports the typing library, the "typing.List" class, and type annotations for function parameters and return types. So I don't see anything here that requires conditional handling.

In any event, the statement `List = "Hello"` makes no sense if you plan to use `List` as a type annotation. This statement is assigning a str instance to `List`, and a str instance cannot be used as an annotation. If mypy is not complaining about this, then I would consider that a bug in mypy.

A static analyzer can't know whether an exception will be raised at runtime, so defining types within a try/except block will inevitably confuse a static type checker. If you really need to conditionalize the behavior based on the Python version (and I still don't understand why you think this is necessary), you should use `if` statements rather than `try` statements. Static type checkers can determine whether an `if` statement evaluates to true or false if you use special expression forms. In particular:
```
if sys.version_info >= (3, 3):
   # Do something
else:
   # Do something else
```

**Comment by AmjadHD** [comment]
I told you it was a proof of concept, mypy does not complain about the code in the library, not my code.
>If you really need to conditionalize the behavior based on the Python version (and I still don't understand why you think this is necessary)

because python 3.3 does not ship with the typing module

why does it behave differently depending on whether the type is imported or not:
1. **Imported**
mytyping.py
```python
try:
    List = "Hello"
except:
    from typing import List
```
foo.py
```python
from mytyping import List

def f(arg: List[str]):
    return arg
```
==> `(type alias) List: Type[List[_T]]`
2. **no imported**
foo.py
```python
try:
	List = "Hello"
except:
	from typing import List

def f(arg: List[str]):
	return arg
```
==> `(variable) List: List | Literal['Hello']`

**Comment by erictraut** [comment]
Ah, I thought the typing module was introduced in 3.3, but I see that it wasn't released until 3.5. There is a [separate installable module](https://pypi.org/project/typing/) that provides backward compatibility with earlier versions.

>why does it behave differently depending on whether the type is imported or not

From the perspective of a static analyzer, you are importing "typing" in both cases. A static analyzer has no idea whether the import will succeed at runtime — or whether any exception will be raised within the try block.

When a static analyzer attempts to determine the type of a symbol (in this case, "List"), it looks at all the places it is assigned. In your examples, it is assigned in two different places: 1) in an "import" statement and 2) in a normal assignment. The import statement provides "List" a type of "Type[List]", and the assignment statement provides "List" a type of "Literal["Hello"]".

**Comment by AmjadHD** [comment]
Yes but the question remains (or i didn't understand), why is it behaving differently ?
in the first scenario (imported), pyright infers `(type alias) List: Type[List[_T]]`, in the second `(variable) List: List | Literal['Hello']`

**Comment by erictraut** [comment]
It behaves differently because of the way Pyright evaluates imported symbols. When a symbol is assigned multiple times within a file and exported from the module, when it is imported by another file, Pyright uses only the last assignment to infer its type.
